### ISSUE TITLE
I want to beep

### ISSUE THREAD

**Issue by mcclure** [ratatui/ratatui#1522]
## Problem

I am implementing a terminal text editor. At times, I must alert the user they did something wrong. Terminals have a builtin "bell" capability that does something system-defined like beep or flash the titlebar. I can activate this by saying `print!("\x07");`.

However, **this is wrong!** I am using ratatui which supports multiple backends. As far as I know all those backends are either targeting VT100 emulators or Microsoft Windows (both of which support `\x07`) but in the future, I could imagine a backend targeting something "weird" like the web DOM.

## Solution

The [Action](https://docs.rs/terminal/latest/terminal/enum.Action.html) Enum should have a new member `Beep` (or possibly `Bell`). In existing backends, this should just `print!("\x07");`.

## Alternatives

Things I think you should not do:
- Name the Action something generic like 'Alert'. It should be unambiguous the new action maps to the "ASCII Bell" terminal verb (which has developed a distinct semantic meaning among programmers unique from the literal bell that used to be attached to the terminals).
- Put additional data such as pitch on the bell/beep verb. Although some methods of beeping can have a pitch, this is the *terminal verb* bell which is just a single signal.

## Additional context

Crossterm has a function [crossterm_treminal_ring_bell](https://docs.rs/libcrossterm/latest/crossterm/fn.crossterm_terminal_ring_bell.html). If ratatui doesn't want to add a bell verb, I could potentially call this directly. However I think it would be nice if my program relied on ratatui for terminal interactions and didn't have to take a dependency on a single backend. (At any rate the crossterm backend can call this function.)

**Comment by joshka** [comment]
Hey there.
https://docs.rs/terminal/latest/terminal/enum.Action.html is not related to Ratatui. We don't use that crate internally anywhere as far as I know. Perhaps you accidentally clicked a wrong link? The `ratatui::Terminal` type is not the same as the `terminal::Terminal` type.

That said, this seems like something reasonable. I wonder if you could maybe rephrase the idea more in terms of what Ratatui could provide to help you beep.

I think what you're saying is that given some frame being drawn, it might be worth being able to tell the render to emit a bell character to the backend. For context, the simplified call path sequence`Terminal::draw`, user code calls methods on `Frame`, `Terminal::draw` sends instructions to backend on what to write.

There's a bit of a caveat. Emitting the bell character would have happen in a way that it doesn't disturb existing cells, which is probably where the main difficulty in implementing this would be (it could also turn out to be fairly simple once someone looks into it deeper - maybe raw mode disables moving the cursor for the bell character IDK off the top of my head.)

Something that you might also consider instead of just the bell is integration with some toast notification system (there are a few of these, I don't know specific libs for rust)

**Comment by mcclure** [comment]
Sorry, I was trying to fit the interface of `ratatui::Terminal<>` and I appear to have messed up on docs.rs. Not sure what I did.

Looking at the [correct docs](https://docs.rs/ratatui/latest/ratatui/struct.Terminal.html), I think my suggestion is that you make `.bell()` or `.beep()` a function on the `ratatui::Terminal<>` struct. I don't think it actually makes sense to merge it with a draw because a bell is something you might want to issue without updating the screen shape in any way.

**Comment by mcclure** [comment]
I will attempt a PR with name `bell()` when I next have time, thanks.

**Comment by mcclure** [comment]
PR is up.

As a note:

> Crossterm has a function [crossterm_treminal_ring_bell](https://docs.rs/libcrossterm/latest/crossterm/fn.crossterm_terminal_ring_bell.html)

I was wrong about this, this is true of the old C-based crossterm crate but not the new pure-Rust one. Apparently I gotta work on my docs.rs navigation skills…!

**Comment by joshka** [comment]
> I was wrong about this, this is true of the old C-based crossterm crate but not the new pure-Rust one. Apparently I gotta work on my docs.rs navigation skills…!

I noticed, but didn't want to make a big deal of it ;)
